### PR TITLE
RFC 239: Policy on LLM assistance in contributions

### PR DESCRIPTION
**Pull request by jugglinmike**
This initial draft takes a maximalist approach (and a permissive stance) to promote a robust and grounded discussion.

[Rendered](https://github.com/bocoup/wpt-rfcs/blob/llm-policy/rfcs/llm_policy.md)

### CODE REVIEW

**Comment by gsnedders** [comment]
Fixes https://github.com/web-platform-tests/rfcs/issues/202

File: rfcs/llm_policy.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+A few examples of policies on LLM use in FOSS contributions:
+
+- permissive
+  - [ghostty/AI_POLICY.md at main · ghostty-org/ghostty](https://github.com/ghostty-org/ghostty/blob/main/AI_POLICY.md)
+  - [Policy about LLM generated code from PRs · Issue #28335 · opencv/opencv](https://github.com/opencv/opencv/issues/28335)
+  - [CONTRIBUTING.md: Guidelines relevant to AI-assisted contributions by gasche · Pull Request #14052 · ocaml/ocaml](https://github.com/ocaml/ocaml/pull/14052)
+  - [LLVM AI Tool Use Policy — LLVM 23.0.0git documentation](https://llvm.org/docs/AIToolPolicy.html)
+- prohibitive
+  - [Code of Conduct ⚡ Zig Programming Language](https://ziglang.org/code-of-conduct/#strict-no-llm-no-ai-policy)
+  - [Getting Started - The Servo Book](https://book.servo.org/contributing/getting-started.html#ai-contributions)
```
Comment (gsnedders) [inline rfcs/llm_policy.md:31]: I feel like there's three others which are notably relevant here: [Chromium's](https://chromium.googlesource.com/chromium/src/+/4f44016dfbd4fcd890694c00d7f9ec6dcefe4955/agents/ai_policy.md), and [Firefox's](https://github.com/mozilla-firefox/firefox/blob/1f7030c8de8f2b349c7d91d7b5a3253c109a1cc1/docs/contributing/ai-coding.md), given they are two of the five repos which have approval to land changes in WPT without further review. (WebKit and Test262 do not currently have policies — [TC39's](https://github.com/tc39/how-we-work/blob/68d1201c1afb4a4155d1ff572d41bb8028f82b3e/AI_POLICY.md) explicitly does not apply to code.)

File: rfcs/llm_policy.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+>
+> #### Disclosure
+>
+> Contributions that contain substantial amounts of tool-generated content must
+> be labeled as such.
+>
+> #### Attribution
+>
+> Commits generated entirely by an LLM must be attributed to the LLM in the
+> "Author" field.
```
Comment (gsnedders) [inline rfcs/llm_policy.md]: This feels problematic. If we attribute a PR to Claude, Gemini, or OpenAI's GPT, if I try and contact the author… well, I don't think Anthropic, Google, or OpenAI are going to be very helpful?

Both Chromium and Firefox's policies are crystal clear that humans are still the authors and must self-review before submitting.

Therefore, when there's still a human very much in the loop who is required to self-review, it does not seem reasonable to consider the LLM the author — and the Chromium policy is explicit that, "Authors must attest that the code they submit is their original creation, regardless of whether AI tooling was used".

File: rfcs/llm_policy.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+## Details
+
+Proposed text:
+
+> ### For Individual Contributors
+>
+> #### Disclosure
+>
+> Contributions that contain substantial amounts of tool-generated content must
+> be labeled as such.
```
Comment (gsnedders) [inline rfcs/llm_policy.md]: Neither Chromium nor Firefox require this today, and it's entirely plausible we've already had commits land into WPT via exports which don't meet this bar.

That said, Chromium's policy here is currently:

> To aid reviewers, authors should flag areas that they are not confident about that had AI assistance. 

This is maybe a weaker form, and hopefully something more in line with existing contributions.

File: rfcs/llm_policy.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+> prepared to engage in technical discussion regarding those changes.
+>
+> ### For Trusted External Review
+>
+> Some external projects conduct review which the WPT maintainers recognize as
+> authoritative. From rendering engines like Gecko to dedicated test suites
+> like WASM, patches merged in these projects are incorporated into WPT without
+> further review. The policy outlined by this document does not apply to these
+> contributions; the external projects are trusted to determine their own
+> mechanisms for quality assurance.
```
Comment (gsnedders) [inline rfcs/llm_policy.md]: This feels like it should probably be at least in part in another RFC that tries to define our existing policies?

As far as I'm aware, there's currently five repos which have approval to incorporate based on downstream review — Chromium, Firefox, Servo, Test262, and WebKit.

My understanding of the unwritten policy is we trust downstream reviewers; I can't even find the various places where we've elucidated parts of the policy over the years.

File: rfcs/llm_policy.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  - [Policy about LLM generated code from PRs · Issue #28335 · opencv/opencv](https://github.com/opencv/opencv/issues/28335)
+  - [CONTRIBUTING.md: Guidelines relevant to AI-assisted contributions by gasche · Pull Request #14052 · ocaml/ocaml](https://github.com/ocaml/ocaml/pull/14052)
+  - [LLVM AI Tool Use Policy — LLVM 23.0.0git documentation](https://llvm.org/docs/AIToolPolicy.html)
+- prohibitive
+  - [Code of Conduct ⚡ Zig Programming Language](https://ziglang.org/code-of-conduct/#strict-no-llm-no-ai-policy)
+  - [Getting Started - The Servo Book](https://book.servo.org/contributing/getting-started.html#ai-contributions)
+
+## Details
+
+Proposed text:
```
Comment (gsnedders) [inline rfcs/llm_policy.md]: Proposed for where?

File: rfcs/llm_policy.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+> ### For Trusted External Review
+>
+> Some external projects conduct review which the WPT maintainers recognize as
+> authoritative. From rendering engines like Gecko to dedicated test suites
+> like WASM, patches merged in these projects are incorporated into WPT without
+> further review. The policy outlined by this document does not apply to these
+> contributions; the external projects are trusted to determine their own
+> mechanisms for quality assurance.
+
+## Risks
```
Comment (gsnedders) [inline rfcs/llm_policy.md:71]: I think it's worthwhile including at least a few more technical risks:

1. Contributions of tests generated by an LLM closely looking at a specific implementation's code, matching that implementation, rather than the spec. (This is, of course, already an issue — but could inevitably become more of a problem if we get more, larger contributions.)
2. Contributions not matching the spec at all. I've seen this mostly with trying to generate tests to assert ordering of things which end of using [HTML's parallelism](https://html.spec.whatwg.org/multipage/infrastructure.html#parallelism) and [HTML's event loops](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop); that case is especially annoying because it can lead to flaky tests.

**Review by gsnedders — CHANGES_REQUESTED** [review]
Thanks for trying to tackle this!

I think we should also write something about use of LLMs for review, if nothing else.

File: rfcs/llm_policy.md (hunk 2 above)
Comment (tabatkins) [reply rfcs/llm_policy.md]: Strong agree on this point for all the reasons you give. The Author field's purpose is to give a contact for problems/question, not assign blame. Listing an LLM is worthless there.

And also, yeah, assigning authorship to an LLM is abrogating your responsibility as an engineer to commit useful code that you understand.

**Comment by martinthomson** [comment]
In terms of a survey, [this](https://redmonk.com/kholterhoff/2026/02/26/generative-ai-policy-landscape-in-open-source/) is a very useful resource.  It's fairly comprehensive and it provides a useful rubric for classification across quality, copyright, and ethics.

File: rfcs/llm_policy.md (hunk 2 above)
Comment (jgraham) [reply rfcs/llm_policy.md]: +1 Commits that attribute authorship to a tool rather than the person using the tool is something that the policy should explicitly ban, not require.

**Review by jgraham — CHANGES_REQUESTED** [review]
My view is that the overall policy should be something like:
* The use of LLMs as tools to help author commits to this repository is allowed
* All contributions must be attributed to a person. That person is expected to fully understand and take responsibility for their contribution, irrespective of any tooling used.
* If LLMs are used as a significant input to a commit, authors are encouraged to include details about how they were used as part of the commit message, in order to help review and future understanding of the code (similar to the way that a commit created using a script might include details of the script).
* People who habitually submit low quality PRs, PRs they don't understand, or otherwise make contributions that are deemed to waste the time of other project contributors or downstream users of the project may be banned.

**Comment by jgraham** [comment]
https://github.com/BurntSushi/ripgrep/blob/master/AI_POLICY.md is also interesting, in particular the "no AI in issues or issue comments" part, with a specific carve out for use cases like translation.

**Comment by DanielRyanSmith** [comment]
I think I'm mostly in agreement with the discussions here so far. I do have one scenario where I think the policy could be expanded on.

> All contributions must be attributed to a person. That person is expected to fully understand and take responsibility for their contribution, irrespective of any tooling used.

I think it would be reasonable to allow a test to be proposed or generated, and although the original author does not have a strong fundamental understanding of the change, it would be feasible to have that change reviewed thoroughly by a domain expert before submission.

For example, using the experimental tool [WPT-Gen](https://github.com/GoogleChromeLabs/wpt-gen) can uncover potential gaps in test coverage reading the spec, and then generate tests to cover those gaps. Hypothetically, we could occasionally run this automatically for certain features. If we have domain expert reviewers who are willing to thoroughly review generated tests, it seems like it would be useful to have an option in the LLM policy to allow for this. Naturally, this does put a lot of burden on the already thankless job of being a reviewer, but it still seems like a viable contribution avenue.

I imagine this will not be terribly common, but it is something that doesn't sound unrealistic.

**Comment by tabatkins** [comment]
As a reviewer, I am not interested in reviewing tests that the author does not understand. There are a lot more test authors than test reviewers. Someone using wpt-gen to create tests should then review those tests and make sure they're doing reasonable things before tossing them at a reviewer. (Generally speaking, I'm not happy with things that further unbalance the effort expectations in the submitter/reviewer dynamic. I'm happy to do thorough review *to teach a person* how to handle an area; I am not happy to do the same to approve/reject a large batch of unreviewed tests when the effort will not meaningfully improve future interactions.))

**Comment by jgraham** [comment]
To expand a bit on my proposal for issue/PR comments:
* The project may choose to use tooling that produces LLM-generated comments. Such comments must always be explicitly associated with a bot account. The core team gets final discretion about which bots / tooling are permitted.
* Human authored comments should not include LLM-generated content in the main text; any such content must be clearly labelled and placed inside a `<details>` element.

**Comment by DanielRyanSmith** [comment]
> As a reviewer, I am not interested in reviewing tests that the author does not understand. There are a lot more test authors than test reviewers. Someone using wpt-gen to create tests should then review those tests and make sure they're doing reasonable things before tossing them at a reviewer. (Generally speaking, I'm not happy with things that further unbalance the effort expectations in the submitter/reviewer dynamic. I'm happy to do thorough review *to teach a person* how to handle an area; I am not happy to do the same to approve/reject a large batch of unreviewed tests when the effort will not meaningfully improve future interactions.))

I don't disagree with any of this, and I should clarify that I'm not advocating for reviewers now being responsible for reviewing AI-generated PRs that authors do not understand - that would absolutely not be beneficial to the project.

I do think that there's a scenario where someone is working to identify and fill gaps in test coverage using LLMs for specific features, and if that person has already interacted with a specific reviewer who is in agreement and willing to review these tests, then that should not be disbarred from contribution if these tests are being reviewed by a competent reviewer.

This is mostly a corner case, but it is something that I wanted to bring up as well.

File: rfcs/llm_policy.md (hunk 6 above)
Comment (jugglinmike) [reply rfcs/llm_policy.md:71]: To me, those look like specific failure modes of LLMs. I think they'd be more helpful to elucidate what we mean by "low-value contributions" rather than as additional risks that are distinct from low-value contributions. I've just pushed a commit incorporating them in that way.

**Comment by jugglinmike** [comment]
Thanks, folks! During today's meeting, I offered to incorporate all the feedback that hasn't received explicit pushback on this discussion thread. While carrying that out, I've found one small conflict between the suggestions above.

@gsnedders [gestured toward](https://github.com/web-platform-tests/rfcs/pull/239#discussion_r3284469366) a statement in the Chromium project's policy:

> That said, Chromium's policy here is currently:
>
> > To aid reviewers, authors should flag areas that they are not confident about that had AI assistance.
>
> This is maybe a weaker form, and hopefully something more in line with existing contributions.

However, @jgraham [later made a suggestion](https://github.com/web-platform-tests/rfcs/pull/239#pullrequestreview-4380258465) which undermines that text:

> - All contributions must be attributed to a person. That person is expected to fully understand and take responsibility for their contribution, irrespective of any tooling used.
> - If LLMs are used as a significant input to a commit, authors are encouraged to include details about how they were used as part of the commit message, in order to help review and future understanding of the code (similar to the way that a commit created using a script might include details of the script).

If a person must "fully understand and take responsibility for their contribution," then there doesn't seem to be a need to guide folks who "are not confident" about their work. I've incorporated the part from @jgraham's exclusively because @gsnedders has not pushed back on it.

**Comment by gsnedders** [comment]
> I do think that there's a scenario where someone is working to identify and fill gaps in test coverage using LLMs for specific features, and if that person has already interacted with a specific reviewer who is in agreement and willing to review these tests, then that should not be disbarred from contribution if these tests are being reviewed by a competent reviewer.

I think there's two meaningful cases here:

1. The specific reviewer doesn't have access to the same models as the other contributor, and
2. The specific reviewer is delegating the task of actually generating the tests to the other contributor.

In either case, I think I'd rather push the other contributor to provide the tests to the specific reviewer in some other way — e.g., by pushing the branch to their fork without opening a PR so the specific reviewer can take over and do any initial fixup work before opening a PR.

I, like others here, don't want to deal with PRs that the author hasn't themselves reviewed.

> If a person must "fully understand and take responsibility for their contribution," then there doesn't seem to be a need to guide folks who "are not confident" about their work. I've incorporated the part from @jgraham's exclusively because @gsnedders has not pushed back on it.

While I recognise @jgraham was clearly deriving text from the Firefox policy, I do wonder if "fully understand" is a somewhat unreasonable bar! I think most of us working on the infrastructure have touched things without fully understanding the code?

That said, I don't have a particularly strong opinion here!

File: rfcs/llm_policy.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+> #### Disclosure
+>
+> If LLMs are used as a significant input to a commit, authors are encouraged
+> to include details about how they were used as part of the commit message in
+> order to help review and future understanding of the code.
+>
+> Human-authored code discourse (e.g. issue descriptions, pull request
+> descriptions, and responses to discussion threads) should not include
+> LLM-generated content in the main text; any such content must be clearly
+> labelled and placed inside a `<details>` element.
```
Comment (annevk) [inline rfcs/llm_policy.md]: This feels too onerous. What if people use it for translation? What if they copy-and-pasted a tiny snippet?

File: rfcs/llm_policy.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+>
+> #### Attribution
+>
+> All commits must be attributed to the human who is taking responsibility for
+> them, regardless of LLM use.
+>
+> #### Understanding
+>
+> Every pull request must be initiated by one human. That person must author
```
Comment (annevk) [inline rfcs/llm_policy.md:67]: Should this be at least one? Don't we sometimes get team contributions?

File: rfcs/llm_policy.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+> the pull request description, understand every change proposed, and be
+> prepared to engage in technical discussion regarding those changes.
+>
+> #### Discussion
+>
+> Discussion
```
Comment (annevk) [inline rfcs/llm_policy.md]: Is this meant to be part of the policy?

File: rfcs/llm_policy.md (hunk 7 above)
Comment (jgraham) [reply rfcs/llm_policy.md]: The ripgrep policy linked above has specific text about translation, which after noting its usefulness says the same policy applies: 
> If using AI for translation, we recommend writing in your native language and including the AI translation in a quote block.

That makes a lot of sense to me. If you only auto-translated the content then providing the source language as canonical allows others to redo the translation potentially using a better tool.

The tiny snippet thing feels very edge-casy, but we can try to reword. Are there specific scenarios you have in mind? Personally I'm not sure why someone would take just a few words directly and intermix it with otherwise handwritten text. If it's not substantive (e.g. you're just copying phrasing) then I'd prefer that you didn't do that, if it is substantive (e.g. you're relying on an LLM for a point of fact) I'd prefer that you are explicit about where you got it from (or rewrite it in your own words, which at least suggests that you put a modest amount of thought into it).

I do think there's a case for allowing `<blockquote>` for including shorter sections of LLM output (up to one paragraph).

File: rfcs/llm_policy.md (hunk 8 above)
Comment (jgraham) [reply rfcs/llm_policy.md:67]: All PRs are created by one user account which almost always means one person. Do you have an example of a PR that's attributed to a team rather than an individual?

File: rfcs/llm_policy.md (hunk 7 above)
Comment (annevk) [reply rfcs/llm_policy.md]: E.g., the LLM whipped up a comparison table to illustrate some design trade-offs. It just seems excessive to have a "must" on this.

File: rfcs/llm_policy.md (hunk 8 above)
Comment (annevk) [reply rfcs/llm_policy.md:67]: That part makes sense, but I don't think the human uploading always knows about every line, if they are open sourcing a set of tests for instance that were developed internally over some period of time. E.g., I think some Opera contributions in the past were of this nature.

File: rfcs/llm_policy.md (hunk 7 above)
Comment (jgraham) [reply rfcs/llm_policy.md]: That feels like a case where it's very useful to know that it was LLM output that was copied and pasted.

File: rfcs/llm_policy.md (hunk 8 above)
Comment (jgraham) [reply rfcs/llm_policy.md:67]: That seems like a case where we're trusting a third party repository as a source of commits (as we do for Gecko, Chromium, WebKit and Servo repositories today). It is true the policy there isn't quite made explicit anywhere, but I think we should deal with that separately compared to the policy for direct contributions to this repository.

File: rfcs/llm_policy.md (hunk 7 above)
Comment (annevk) [reply rfcs/llm_policy.md]: I personally don’t really mind where it comes from as long as a human pasted it and knows it’s correct. If they are unsure, then yes.

Being overtly strict doesn’t seem like the right trade-off. Especially since I don’t think we even have that many bad contributions? They first have to figure out how to get reviewed. 😅

File: rfcs/llm_policy.md (hunk 8 above)
Comment (gsnedders) [reply rfcs/llm_policy.md:67]: > Do you have an example of a PR that's attributed to a team rather than an individual?

Maybe irrelevant detail, but you can create PRs whose head points at an organization's fork, though it is still a user account actually opening the PR.

File: rfcs/llm_policy.md (hunk 6 above)
Comment (gsnedders) [reply rfcs/llm_policy.md:71]: While I think the (2) might be reasonable to class as "low value", I think (1) is really a different failure mode, and one that becomes a higher-risk with LLMs (by virtue of sheer volume).

File: rfcs/llm_policy.md (hunk 7 above)
Comment (jugglinmike) [reply rfcs/llm_policy.md]: @annevk I think such an exception would probably have to be a little more precise than "a tiny snippet." That might also help folks engage with the way you're considering risks. What quantity of generated text needs a label?

File: rfcs/llm_policy.md (hunk 8 above)
Comment (jugglinmike) [reply rfcs/llm_policy.md:67]: @annevk If you're satisfied by James' latest response, could you mark this as resolved?

File: rfcs/llm_policy.md (hunk 9 above)
Comment (jugglinmike) [reply rfcs/llm_policy.md]: It is not! I've removed it.

File: rfcs/llm_policy.md (hunk 8 above)
Comment (annevk) [reply rfcs/llm_policy.md:67]: I'm not sure I'm entirely satisfied, but I'm also not sure it matters much. I don't think we have to necessarily have a relationship with a third-party repository for someone to contribute tests from a third-party repository. Like if someone had an independent implementation of the URL standard and they and other contributors created a bunch of tests for it, I would not want to enforce that the person upstreaming those tests can defend and speak to them all themselves.

File: rfcs/llm_policy.md (hunk 7 above)
Comment (annevk) [reply rfcs/llm_policy.md]: If you have a human written summary and want to defend that with a bunch of LLM text then I think it needs a label and it probably helps if it's hidden by default as there's not a lot of care put into it. But if you create a response that stands on its own it's probably fine. I'm not sure we have enough experience to judge one way or another.

**Comment by jugglinmike** [comment]
My understanding is that while we expect this policy may evolve with experience, there are no objections to implementing it in its current form. To that end, @jgraham or @gsnedders: could you formalize your approval so that the acceptance countdown can begin?

File: rfcs/llm_policy.md (hunk 6 above)
Comment (jugglinmike) [reply rfcs/llm_policy.md:71]: I agree that (1) is distinct from (2), but I still don't see how it represents a meaningfully-distinct failure mode of this policy. Both (1) and (2) are second-order risks from defining a policy. They both flow from the premise that any such policy will encourage the use of tools that have these weaknesses. And they both have the same impact on our risk calculation here: we can't technically stop people, but an explicit policy formalizes a social contract.

That's what's keeping me from creating a new sub-section for over-fitting. It's easy to frame that as a separate risk, but it seems like we can only come to the same conclusion as we do for the "low-value contribution" risk: we accept it because we expect people to act in good faith and follow the policy.

File: rfcs/llm_policy.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+> #### Disclosure
+>
+> If LLMs are used as a significant input to a commit, authors are encouraged
+> to include details about how they were used as part of the commit message in
+> order to help review and future understanding of the code.
+>
+> Human-authored code discourse (e.g. issue descriptions, pull request
+> descriptions, and responses to discussion threads) should not include
+> LLM-generated content in the main text; any such content must be clearly
+> labelled and placed inside a `<details>` element.
```
Comment (jgraham) [inline rfcs/llm_policy.md]: I think perhaps we should allow "blockquote" for short snippets (say: up to one paragraph).

**Review by jgraham — APPROVED** [review]

File: rfcs/llm_policy.md (hunk 10 above)
Comment (jugglinmike) [reply rfcs/llm_policy.md]: Done!

**Comment by jgraham** [comment]
@gsnedders are your concerns still blocking here?

**Review by jcscottiii — APPROVED** [review]

File: rfcs/llm_policy.md (hunk 7 above)
Comment (gsnedders) [reply rfcs/llm_policy.md]: I think this might be slightly too strict, but on balance I think having a policy is better than not — let's keep this as is, and we can revisit if people find it onerous.

**Review by gsnedders — APPROVED** [review]
